packages/components/src/Toast/Toast.tsx

```diff
@@ -0,0 +1,123 @@
+import React, { useEffect, useState } from "react";
+import { AnimatePresence, motion } from "framer-motion";
+import { XOR } from "ts-xor";
+import styles from "./Toast.css";
+import { Icon, IconColorNames, IconNames } from "../Icon";
+import { Button } from "../Button";
+
+interface BaseToastProps {
+  readonly variation?: "success" | "error";
+  readonly message: string;
+  readonly id?: number;
+}
+
+interface ActionToastProps extends BaseToastProps {
+  action(): void;
+  actionLabel: string;
+}
+
+export type ToastProps = XOR<BaseToastProps, ActionToastProps>;
+type ToastPropsInternal = Omit<ToastProps, "id">;
+
+export interface ToastRef {
+  add(props: ToastProps): void;
+}
+
+interface Icon {
+  name: IconNames;
+  color: IconColorNames;
+}
+
+export function Toast({
+  message,
+  variation,
+  action,
+  actionLabel,
+}: ToastPropsInternal) {
+  const [visible, setVisible] = useState(true);
+  const icon = getIcon();
+  let timer: NodeJS.Timeout;
+
+  useEffect(() => startTimer(), []);
+
+  return (
+    <AnimatePresence>
+      {visible && (
+        <motion.div
+          className={styles.toast}
+          initial={{ opacity: 0, scale: 0.8 }}
+          animate={{ opacity: 1, scale: 1 }}
+          onMouseEnter={() => stopTimer()}
+          onMouseLeave={() => startTimer()}
+          exit={{
+            opacity: 0,
+            scale: 0.8,
+            height: 0,
+            transition: { duration: 0.4 },
+          }}
+        >
+          <div className={styles.slice}>
+            {icon && (
+              <div className={styles.icon}>
+                <Icon color={icon.color} name={icon.name} />
+              </div>
+            )}
+
+            <div className={styles.message}>{message}</div>
+
+            {action && (
+              <Button
+                label={`${actionLabel}`}
+                onClick={action}
+                type="tertiary"
+              />
+            )}
+
+            <div className={styles.button}>
+              <Button
+                icon="remove"
+                ariaLabel={"Hide Notification"}
+                onClick={handleToastClose}
+                type="tertiary"
+                variation="cancel"
+              />
+            </div>
+          </div>
+        </motion.div>
+      )}
+    </AnimatePresence>
+  );
+
+  function startTimer() {
+    timer = setTimeout(() => handleToastClose(), getTimeout());
+  }
+
+  function stopTimer() {
+    clearTimeout(timer);
+  }
+
+  function handleToastClose() {
+    setVisible(false);
+  }
+
+  function getTimeout() {
+    const min = 2000;
+    const max = 5000;
+    const time = message.length * 100;
+
+    if (time < min) return min;
+    if (time > max) return max;
+    return time;
```

Comment:
I wonder if `Math.min` and `Math.max` would be better here?

Reply:
@seanhealy I don't think that there is anything to gain from changing this. I think that the readability of what is here is more simple.